add rand dispatch for tuple types

adienes opened the pull request:
as suggested by https://github.com/JuliaLang/julia/issues/50236 @rfourquet

adienes committed:
add rand dispatch for tuple types

adienes commented:
I doubt those build failures are related to this change

rfourquet commented:
I'm not a fan of using broadcast for the implementation. Performance-wise, we should try to not be too far-off from RandomExtensions.jl:
```
# this PR
julia> @btime rand(Tuple{Bool,Int})
  1.310 μs (20 allocations: 744 bytes)

# RandomExtensions
julia> @btime rand(Tuple{Bool,Int})
  7.722 ns (0 allocations: 0 bytes)
```

adienes committed:
remove allocating broadcast

adienes commented:
ok, should match performance now (at least locally for me)

stevengj commented on stdlib/Random/src/generation.jl:
```diff
@@ -167,6 +167,11 @@ function rand(r::AbstractRNG, ::SamplerType{T}) where {T<:AbstractChar}
+
+function rand(r::AbstractRNG, ::SamplerType{T}) where {T<:Tuple}
+    ntuple(i -> rand(r, T.parameters[i]), length(T.parameters))
+end
```
Is this guaranteed to be type stable?

I wonder if we should use a generated function instead.

adienes replied:
good point. I see `ntuple` allocates past `10` parameters, I can try the `Val{N}` version. also I didn't check that this works with `Vararg` so there might be a little more work to do there

adienes replied:
I think this is basically what the `Val` version does, under an `if @generated` branch

```
    if @generated
        :(@ntuple $N i -> f(i))
```

where
```
macro ntuple(N::Int, ex)
    vars = Any[ inlineanonymous(ex,i) for i = 1:N ]
    Expr(:escape, Expr(:tuple, vars...))
end
```

I found in the `Metaprogramming` section the guidance `This [optionally @generated] style is preferred, since it allows the compiler to make more decisions and compile
programs in more ways`, hence my option to use the existing `ntuple(, ::Val)`

Anyway, I tried both versions on a wide tuple (~200 parameters) and they had nearly identical performance within a few nanos :)

vtjnash replied:
This should not access `.parameters` (which a private, possibly unstable, API), but rather map over the `fieldtype(T, i)`

adienes replied:
@vtjnash how about `ntuple(i -> rand(r, fieldtype(T, i)), Val(fieldcount(T)))` ?

martinholters replied:
`map(t -> rand(r, t), fieldtypes(T))`?

stevengj replied:
or just `rand.(r, fieldtypes(T))`?

adienes replied:
thank you both for the suggestions. both of those alternate implementations are indeed easier to read, but unfortunately they seem to have worse performance on wide & nested types

```
### broadcasting
julia> @btime rand(Tuple{Bool, Char, Vararg{Tuple{Int, Float64, Tuple{Char, Bool}}, 80}})
  46.417 μs (486 allocations: 12.77 KiB)


### map
julia> @btime rand(Tuple{Bool, Char, Vararg{Tuple{Int, Float64, Tuple{Char, Bool}}, 80}})
  10.208 μs (85 allocations: 6.59 KiB)


### this PR
julia> @btime rand(Tuple{Bool, Char, Vararg{Tuple{Int, Float64, Tuple{Char, Bool}}, 80}})
  1.212 μs (80 allocations: 2.50 KiB)
```

stevengj commented on stdlib/Random/test/runtests.jl:
```diff
@@ -333,7 +333,7 @@ end
     ftypes = [Float16, Float32, Float64, FakeFloat64, BigFloat]
     cftypes = [ComplexF16, ComplexF32, ComplexF64, ftypes...]
-    types = [Bool, Char, BigFloat, Base.BitInteger_types..., ftypes...]
+    types = [Bool, Char, BigFloat, Tuple{Bool, Tuple{Int, Char}}, Base.BitInteger_types..., ftypes...]
```
Would be good to add a test for `@inferred` to make sure it is type-stable with a tuple-type argument.

adienes committed:
use opt. generated ntuple & add @inferred test

adienes committed:
remove access to internal properties

adienes committed:
add rational rand dispatch

adienes committed:
add to news

adienes commented:
given that this won't be in 1.10 either way and can sit on master for a while, I added in support also for sampling from `Rational` type and updated the docstring of `rand`; would appreciate a review 🙂

rfourquet commented:
I suggest removing the `Rational` part from this PR, which would almost surely prevent merging. See #25993 and #46060 for other attempts, if you want to purse this idea, a separate PR is warranted.

adienes committed:
remove rational rand

adienes commented:
> I suggest removing the `Rational` part from this PR, which would almost surely prevent merging. See #25993 and #46060 for other attempts, if you want to pursue this idea, a separate PR is warranted.

fair enough. I should have guessed that would be contentious; I've removed the `Rational` piece

rfourquet commented:
marking for triage as suggested [here](https://github.com/JuliaLang/julia/pull/28705#issuecomment-1643868379)

rfourquet commented:
I'm in favor of merging this PR as is, although there is room for optimization for some niche cases. Essentially, this implementation doesn't use the 2-staged random generation, i.e. the fact that in a call to `rand`, which first create a sampler, and then call rand on the sampler; this means that for cases (rare, in particular when the distribution is a type like here, a tuple type) where the sampler creation can factor out some non-negligible computation when a bunch a random values are about to be generated (like in an array). For reference, consider the following made-up example:

```
struct ZZ{N}
    x::Char
end

Random.rand(rng::AbstractRNG, sp::Random.SamplerSimple{ZZ{N}}) where {N} = ZZ{N}(rand(rng, sp.data))

function Random.Sampler(::Type{RNG}, ::Type{ZZ{N}}, nn::Random.Repetition) where {N, RNG<:AbstractRNG}
    Random.SamplerSimple(ZZ{N}('a'), Random.Sampler(RNG, String(N)))
    # note: passing `ZZ{N}('a')` is somewhat artificial here, using SamplerTag would be better but i believe it's undocumented
end
```
Then creating an array of such values wrapped in a tuple has significant overhead compared to raw values, unlike the no-overhead approach of RandomExtensions:
``` 
julia> @btime rand(ZZ{:asd}, 1000);
  2.712 μs (3 allocations: 4.16 KiB)

julia> @btime rand(Tuple{ZZ{:asd}}, 1000); # This PR
  34.675 μs (2001 allocations: 105.62 KiB)

julia> @btime rand(Tuple{ZZ{:asd}}, 1000); # Using instead `RandomExtensions`
  2.678 μs (3 allocations: 4.16 KiB)
```
But I think this optimization can be added later if anyone finds the motivation.

LilithHafner commented:
The possibility of `rand(Tuple{1:4, Int}, 10)` did not make triage smile, nor did traige have anything authoritative to say about that possibility.

adienes commented:
thank you for the comments.

I am happy to give it a shot to get the performance of long arrays of custom struct types performant compared to `RandomExtensions.jl`, but I may end up needing help on that

also, I did not know `Tuple{1:4, Int}` was even allowed 😅, don't all the parameters have to be types?

LilithHafner commented:
> don't all the parameters have to be types?

Nope! Types and `isbitstype` values are also allowed. This is useful because it allows `Array{Float, 2}` to be a two-dimensional array with dimensionality known at compile time. A unit range would be a reasonable type parameter for a static offset array.

You can also do silly things like `eltype(Vector{:blue}) === :blue`

adienes commented:
implementing like so
```
function Sampler(RNG::Type{<:AbstractRNG}, t::Type{T}, n::Repetition) where {T <: Tuple}
    SamplerTag{t}(
        ntuple(i -> Sampler(RNG, fieldtype(t, i), n), fieldcount(t))
    )
end

function rand(rng::AbstractRNG, sp::SamplerTag{<:T}) where T<:Tuple
    ntuple(i -> rand(rng, sp.data[i]), fieldcount(T))
end
```

matches performance in the `ZZ` case, but regresses by about an order of magnitude in the wide tuple case,
```
@btime rand(Tuple{Bool, Char, Vararg{Tuple{Int, Float64, Tuple{Char, Bool}}, 80}});
```

still working on it

adienes committed:
faster sampler generation

adienes committed:
get rid of implicit `convert`

adienes committed:
fix double backticks

adienes committed:
use Base.tail

adienes commented:
@rfourquet  could you possibly review the latest updates?
in particular, let me know if the extra allocations in the case of the super-wide heterogenous tuple are acceptable. I suspect to remove those the implementation would have to grow significantly in complexity

rfourquet commented on stdlib/Random/src/Random.jl:
```diff
@@ -313,12 +313,26 @@ include("XoshiroSimd.jl")
-  integers (this is not applicable to [`BigInt`](@ref)), to ``[0, 1)`` for floating
-  point numbers and to ``[0, 1)+i[0, 1)`` for complex floating point numbers;
+
+* a type (and corresponding domain) from the list below
+
+  + `typemin(S):typemax(S)` for concrete integer types (excepting [`BigInt`](@ref) which is not supported)
```
i find this wording confusing: 

> a type (and corresponding domain) from the list below

this implies that the list below will have types as items, but items are instead the domain...

adienes replied:
how about reversing the order of the items? so that would be e.g.

> a type from the list below, which samples uniformly over that type's corresponding domain

* concrete integer types `T` sample from `typemin(T):typemax(T)` (excepting `BigInt`...)
* concrete floating point types sample from `[0, 1)`
* concrete complex types `Complex{T}` sample from `rand(T) + rand(T) * im`
* `<:AbstractChar` types sample from all valid Unicode scalars

rfourquet commented on stdlib/Random/src/generation.jl:
```diff
@@ -167,6 +167,20 @@ function rand(r::AbstractRNG, ::SamplerType{T}) where {T<:AbstractChar}
 
+### random tuples
+
+function Sampler(RNG::Type{<:AbstractRNG}, t::Type{T}, n::Repetition) where {T <: Tuple}
```
Unless someone is assertive that it will be equivalent performance wise, I'd rather have this:
```suggestion
function Sampler(::Type{RNG}, ::Type{T}, n::Repetition) where {T <: Tuple} where {RNG<:AbstractRNG}
```
And use `T` instead of `t` below. It's been a long while since i checked, but i remember at least once where this would prevent type instabilities.

rfourquet commented:
> let me know if the extra allocations in the case of the super-wide heterogenous tuple are acceptable

I think it would be fair to merge this so that things move forward, but it's up to triage...
 
If this version is merged, I will have to update RandomExtensions before the 1.11 release, and in the process I will probably be able to extract from it a reasonable (not too complex) implementation which could supersede this PR; then I hope there won't be much opposition to such an update.

An alternative would be for you or me to directly pull RandomExtensions's implementation.

adienes commented:
do note that since this was last evaluated on triage, the update has now matched performance with `RandomExtensions` on all cases except for the extremely wide tuple case. while I am not an expert on either implementation, my understanding of the reasoning is that `Base.ntuple` basically rewrites/unrolls up to length `10`, while `RandomExtensions` does a lot more codegen in the `@make` macros

given that this is kind of a more fundamental limitation to the performance of `Tuple` in Base rather than anything to do with `rand`, I'm inclined to think that the difference in performance is acceptable. however, I would completely understand if its desired to leave no performance on the table in these wide cases --- that being said, I'm not sure I have the requisite familiarity to pull the code over. I'm happy to attempt with some guidance, and I don't mean to put more work on your plate, but going that route it will probably be best if I let you handle it :)

LilithHafner commented:
Does this support `rand(Tuple{Int, 1:4})`? Does this support `rand(Tuple{Int, Vararg{Int}})` and if so, what are the semantics? Whether or not these are supported, there should be tests for them.

adienes commented:
intentionally not supported to both. I'll go ahead and add tests and make the docs more clear on that

LilithHafner commented:
Once tests and docs are in order, this is good to merge!

adienes committed:
update docs, apply code review, add tests

adienes commented:
green!

adienes commented on stdlib/Random/src/Random.jl:
```diff
@@ -313,12 +313,27 @@ include("XoshiroSimd.jl")
+
+  + concrete floating point types sample from `[0, 1)`
+
+  + concrete complex types `Complex{T}` sample from `rand(T) + rand(T) * im`
```
how about
> concrete complex types `Complex{T}` when `T` is a sampleable type take their real and imaginary components independently from the set of values corresponding to `T`, but are not supported If `T` is not sampleable.

adienes committed:
Update Random.jl complex domain docstring

adienes committed:
Update Random.jl complex docstring

rfourquet committed:
Merge branch 'master' into rand_tuple_type

rfourquet committed:
fix test caused by merge with master
